devito/function.py

```diff
@@ -348,18 +349,20 @@ def argument_values(self, alias=None, **kwargs):
         # Add value override for own data if it is provided
         if self.name in kwargs:
             new = kwargs.pop(self.name)
+            if len(new.shape) != self.ndim:
+                raise ValueError("Array shape %s does not match" % (new.shape, ) +
+                                 "dimensions %s" % (self.indices, ))
             if isinstance(new, TensorFunction):
                 # Set new values and re-derive defaults
                 values[key] = new._data_buffer
                 values.update(new.argument_defaults(alias=key).reduce_all())
             else:
                 # We've been provided a pure-data replacement (array)
                 values[key] = new
-                # TODO: Re-derive defaults from shape of array
+                # Add value overrides for all associated dimensions
+                for i, s, o in zip(self.indices, new.shape, self.staggered):
```

Comment:
Shouldn't this go directly under the first ``if self.name in kwargs``?

Reply:
No, this one is only necessary if an array is passed as `new. argument_defaults` will take care of indices and shape for a `TensorFunction`.